Makefile flow

SgtCoDFish opened the pull request:
## Summary

This PR provides a `make`-based flow for building all cert-manager artifacts, with the intention of replacing bazel for deployments, with minimal changes. This isn't intended to fully remove bazel _today_, but instead to remove it for the development edit-build-run workflow and for cert-manager deployments, which are currently handled by `cmrel`.

## Help for Reviewers

If you follow both "Comparison Against Old Releases" and "Building Everything + Signing" you should be able to easily compare an old bazel release and a "new" make release.

### Comparison Against Old Releases

The goal here is that `bin/release` should contain everything needed for `cmrel publish` to be able to publish a release based on its contents, with minimal changes. 

To compare `bin/release` with a previous release, the following may be helpful to check out an existing cert-manager release locally:

```console
$ mkdir -p /tmp/maker/bzl
$ gsutil -m cp -r gs://cert-manager-release/stage/gcb/release/v1.6.0-beta.0-49914a057b39c887be0974c4657c095bd7724bc7 /tmp/maker/bzl/
$ ls /tmp/maker/bzl/v1.6.0-beta.0-49914a057b39c887be0974c4657c095bd7724bc7
...
```

### Building Everything + Signing

Creating a full staged release includes signing the helm chart, which requires a GCP KMS key. I have a test one in my project and I can allowlist you to use it if you ask. Then, you should be able to run the following:

```console
$ export CMREL_KEY="projects/jetstack-ashley-davis/locations/global/keyRings/testring/cryptoKeys/sha512key/cryptoKeyVersions/1"
# note you need to be allowlisted to sign using this test key! ask if you want that, or else you can create your own.
$ make -j8 -f make/Makefile staged-release
```

## High Level Structure

The intention of this PR is that _everything_ we need to stage a deployment can be built with exactly one command, including generating signatures. This build step should also be trivially run in parallel using make's `-jN` option.

```console
$ make -j8 -f make/Makefile <target>
# lots of output...
```

The above command, starting from a clean repo with no `bin/` dir but with a populated go build and module cache takes under 2 minutes to completely build and sign every artifact for a release when run on a developer laptop. The 2 minutes includes downloading and installing tools as dependencies (helm, cosign, etc).

## Artifacts

Everything that's built is placed under `bin/`. Crucially, using `make -f make/Makefile staged-release` will create `bin/release` containing the same artifacts which bazel built, in the same format. In other words, `bin/release` should contain everything required to stage a complete release and side effect of this is that it should **obsolete most of the logic in `cmrel stage`** and allow a lot of cleanup there.

After this PR merges, `cmrel stage` will essentially become "clone repo, run make staged-release, upload to a bucket".

## Notes / Differences from Bazel Artifacts

Lines prefixed with a :warning: denote functional changes which could affect others. Lines without a warning sign should be non-functional changes (e.g. a comment change)

### Versioning

- If HEAD is not a tag, a version like `v1.6.0-67-gf4e2b2c3c` will be generated using `git describe --tags`. This should uniquely identify practically all development builds.
   - The format is `v<last-tag>-<N>-<commit-sha>` where `N` is the number of commits since `<last-tag>`
   - This format is directly from git.
- If HEAD is a tag, the tag will be used for the version.

### Windows Client Binaries

- We add a zip file for windows client binaries since tar.gz is less common on Windows
- :warning: All windows client binaries now have a ".exe" extension

### Helm Charts

- We now write helm chart filenames using the correct version _in_ the filename, rather than just naming the chart `cert-manager.tgz`. This more closely matches how helm expects chart filenames.
- :warning: Helm charts are now packaged into `cert-manager-manifests.tar.gz` with their version in their filename
- The helm chart signing being moved into the cert-manager repo means that it's possible to embed the artifacthub `signKey` annotation directly into the actual helm chart, which should improve the ArtifactHub UX. Previously this annotation was only added in our charts repo's `index.yaml` where it seemed to largely be ignored.

### Signing

- The make flow has support for signing helm charts (using `cmrel` as an installed standalone tool).
- Includes a commented-out template for generating and signing a SHA256SUMS file automatically generated by the build process.
- :warning: While not a breaking change, it's worth highlighting that the signing key information is currently hardcoded in a single file - `deploy/charts/cert-manager/signkey_annotation.txt`. This can obviously be improved later.

### Containers

- Containers are built with docker by default, but this is configurable through the `CTR` environment variable and both `podman` and `nerdctl` can be used as drop-in replacements.
- :warning: When we build containers, the tag looks like `cert-manager-controller-<arch>:<version>` and doesn't include the registry (`quay.io/...`). This is to avoid creating a race condition if building containers in parallel. This will require a small change to `cmrel publish`, but seems worth to avoid this killing the ability to build containers in parallel.

#### Base Images

- Adds a script for automatically updating base images, both distroless/base and distroless/static across all relevant arches.
- All containers now use a distroless base image whose tag matches their arch (e.g. a s390x base for the s390x cert-manager containers). With bazel all containers used amd64, which never caused a problem with the distroless static images but would break if anyone tried to use dynamic or a debug distroless image on non-amd64 platforms.
- Using distroless/base or distroless/static is now a configurable through a single env var (`BASE_IMAGE_TYPE`)

#### Binary Locations

Background: Bazel adds binaries to containers at a long unusual path, along with a symlink:

```console
# taken from a layer in one of our webhook docker images
$ tree .
app
└── cmd
    └── webhook
        ├── webhook -> /app/cmd/webhook/webhook.runfiles/com_github_jetstack_cert_manager/cmd/webhook/webhook_/webhook
        └── webhook.runfiles
            └── com_github_jetstack_cert_manager
                ├── cmd
                │   └── webhook
                │       └── webhook_
                │           └── webhook
                └── external -> /app/cmd/webhook/webhook.runfiles
```

In other words, the *actual binary* is `/app/cmd/webhook/webhook.runfiles/com_github_jetstack_cert_manager/cmd/webhook/webhook_/webhook` and `/app/cmd/webhook/webhook` exists as a symlink to that binary.

- :warning: The make version only adds the actual binary, this time at `/app/cmd/webhook/webhook`.
- :warning: The make version doesn't add the `external` symlink seen in the tree above, either.

#### Licenses in Containers

- :warning: The `LICENSES` file changed into a short string pointing people at our repo if they want to see licenses.
   - This is because `LICENSES` is 1.4MB on its own, which is _absurdly_ wasteful for a file which nobody ever looks at or cares about.
   - This change alone shaves maybe ~3%-5% off the size of the cert-manager containers, basically for free.

### Static manifests: cert-manager.yaml / cert-manager.crds.yaml

- Some comments are different.
- The static manifest now has the namespace resource at the top of the file rather than after the CRDs

## Release Notes

/kind feature

```release-note
Added a makefile-based build workflow which doesn't depend on bazel
```

irbekrm commented on make/containers.mk:
```diff
@@ -0,0 +1,90 @@
+.PHONY: cert-manager-controller-linux
+cert-manager-controller-linux: bin/containers/cert-manager-controller-linux-amd64.tar.gz bin/containers/cert-manager-controller-linux-arm64.tar.gz bin/containers/cert-manager-controller-linux-s390x.tar.gz bin/containers/cert-manager-controller-linux-ppc64le.tar.gz bin/containers/cert-manager-controller-linux-armv7.tar.gz
+
+bin/containers/cert-manager-controller-linux-%.tar.gz: bin/controller-linux-% make/Containerfile.controller | bin/containers
```
So this will be re-run every time the built binary has changed or the Containerfile has changed. What if TAG, BASE_IMAGE or GOARGS has changed? (For example, with Bazel we use [stamping](https://github.com/jetstack/cert-manager/tree/master/build#stamping) to ensure that targets will re-run i.e if a different image tag is set)

maelvls replied:
Good point. I can't think of any "good" way to rebuild the target on env var changes.

I think it is not a deal-breaker though 😄   It wouldn't be a problem on CI anyways, but maybe it would become problematic locally.

irbekrm reviewed (commented):
I've read through the scripts and I think in general this would be a huge improvement as it makes the build steps a lot less complex and I think that's important both because it will be easier to make changes in the future and because it will be less likely that bugs are introduced or scripts drift due to folks not really understanding them.

I agree with the approach in general, would like to play with these a bit more and think about any possible edge cases around ensuring that dependency etc changes are detected and understand the signing/sha generation a bit better. Also interested how this will fit into our e2e test setup

SgtCoDFish commented on make/Makefile:
```diff
@@ -0,0 +1,59 @@
+GO=go
+CMREL=/home/ashley/workspace/cert-manager-release/bin/cmrel
+CTR=docker
```
> do we expect for this to make it easy to pass another tool, i.e podman?

That's exactly why I did it yeah; my experience in the past has been that they can be used pretty interchangably as long as you're not doing anything too fancy. For our use cases I think it should be simple enough to change between them (and I did use podman a little while testing this makefile stuff)

> Thought: having used ko for internal project...

I totally accept there could be a better tool for this; I tried not to add any new tools here just to limit the size of the change (and because there's a risk that none of this gets used :wink: )

I do think this approach makes it _very_ easy to build alternative images (as you mention) which is a definite win. Honestly because building with this makefile is very quick, I don't think speed is a great motivator here to use something else.

maelvls replied:
> Do we expect for this to make it easy to pass another tool, i.e podman? I guess the required build args might differ though

I have tested with both `nerdctl` and `podman`; it seems like the subset of commands that we are using are compatible with all three (docker, nerdctl and podman) which is great because I default to using nerdctl in a rootless way :+1:

SgtCoDFish commented on make/Makefile:
```diff
@@ -0,0 +1,59 @@
+GO=go
+CMREL=/home/ashley/workspace/cert-manager-release/bin/cmrel
```
NB: this is obviously an artifact of the PR being WIP :)

SgtCoDFish commented on make/Makefile:
```diff
@@ -0,0 +1,59 @@
+VERSION ?= $(shell cat VERSION)
+RELEASE_VERSION ?= $(VERSION)
+
+CMREL_KEY="projects/jetstack-ashley-davis/locations/global/keyRings/testring/cryptoKeys/sha512key/cryptoKeyVersions/1"
+COSIGN_KEY="gcpkms://projects/jetstack-ashley-davis/locations/global/keyRings/testring/cryptoKeys/sha512key/versions/1"
```
NB: this is obviously an artifact of the PR being WIP :)

SgtCoDFish commented on make/Makefile:
```diff
@@ -0,0 +1,59 @@
+	@mkdir -p $@
+
+.PHONY: forcetime
+forcetime: | bin
```
This is sort of copying bazel's timestamping stuff where it tries to be more reproducible by fixing timestamps; e.g. if you pull a cert manager docker image from quay.io and extract it the date on the binary is 01/01/1970. We might not need or want this, but I wanted it to be available if that's something we do want.

```console 
$ ls -lah app/cmd/controller/controller.runfiles/com_github_jetstack_cert_manager/cmd/controller/controller_/controller
-r-xr-xr-x 1 ashley ashley 58M Jan  1  1970 app/cmd/controller/controller.runfiles/com_github_jetstack_cert_manager/cmd/controller/controller_/controller
```

SgtCoDFish commented on make/server.mk:
```diff
@@ -0,0 +1,74 @@
+.PHONY: server-binaries
+server-binaries: controller acmesolver webhook cainjector
```
On a developer laptop, we don't get much because the server binaries obviously need to be run in k8s; the idea for these targets is that they could be run in build containers and copied across in a multi-stage container build.

(put another way, this is more about defining the command required to build the container rather than it being a target we'd use for dev, although it can be useful as a command to build + check that the thing builds successfully)

maelvls replied:
I think it is totally OK to build it outside of a container since we have no dependency on the host (no dynamically linked libraries, not even `libc.so`, and no OS-specific paths embedded thanks to `-trimpath`). The only dependency is Go itself.

At first, I thought "this is going to make the build unreproducible!!". The more I think about it, the more I think I'm OK with that since it allows for a massive speed-up while developing.

It would be nice if there was a target `binary` that would only build linux-amd64 though since that's what (probably) most people will want while developing.

SgtCoDFish replied:
I think ideally we won't be building them locally and copying in, if only because it'll violate SLSA3's requirement for an [ephemeral environment](https://slsa.dev/requirements#ephemeral-environment) - plus there could be differences in go versions on dev laptops. It's _totally_ fine for quick development work, but the main reason this target is defined is so that it can be invoked inside a Containerfile. Any build step run anywhere should have a make target for it.

> It would be nice if there was a target binary that would only build linux-amd64 though since that's what (probably) most people will want while developing.

I think this is served by the `bin/controller-linux-amd64` target right? E.g. `make -f make/Makefile bin/controller-linux-amd64`

SgtCoDFish commented on hack/latest_base_images.sh:
```diff
@@ -0,0 +1,30 @@
+#!/usr/bin/env bash
+
```
Thank you, and thank you for reviewing :heart:

maelvls commented:
This looks great! I tried using nerdctl (containerd and buildkit both running rootless)

```
make -f make/Makefile CTR=nerdctl all-containers
```

It works as expected.

I think we should go on and have this PR merged and use that when releasing for 1.7.0-alpha.0.

SgtCoDFish commented on hack/concat-yaml.sh:
```diff
@@ -0,0 +1,29 @@
+#!/usr/bin/env bash
```
(missed this comment when replying previously, sorry!)

This is for concatenating crds together, which is currently done in [deploy/crds/BUILD.bazel](https://github.com/jetstack/cert-manager/blob/master/deploy/crds/BUILD.bazel#L14-L19) (and it's done unsafely in bazel because it doesn't check for `%YAML` lines)

This is used in `bin/scratch/crds/all-crds.yaml` defined in `make/helm.mk` for now.

I'm not actually sure how if we could use `helm template` for that - but I'm sure there are better tools available!

JoshVanL replied:
Same, re `hack/util`.

SgtCoDFish commented on .dockerignore:
```diff
@@ -0,0 +1,4 @@
+bin/*
+
+!bin/server/**
+!bin/scratch/cert-manager.license
```
note: this means "ignore everything in `bin` except these two paths", which are the only ones currently consumed by the Containerfile

SgtCoDFish commented on deploy/charts/cert-manager/signkey_annotation.txt:
```diff
@@ -0,0 +1,2 @@
+fingerprint: 1020CF3C033D4F35BAE1C19E1226061C665DF13E
+url: https://cert-manager.io/public-keys/cert-manager-keyring-2021-09-20-1020CF3C033D4F35BAE1C19E1226061C665DF13E.gpg
```
note: as mentioned in the PR description this isn't ideal, but it can easily be updated later

SgtCoDFish commented on deploy/manifests/namespace.yaml:
```diff
@@ -0,0 +1,4 @@
+apiVersion: v1
+kind: Namespace
+metadata:
+  name: cert-manager
```
note: the other similar file has `---` in it; a new file is the easiest backwards compatible way I could think of here. Eventually, the other file will go away with bazel.

maelvls commented on make/containers.mk:
```diff
@@ -0,0 +1,123 @@
+bin/containers/cert-manager-controller-linux-amd64.tar.gz bin/containers/cert-manager-controller-linux-arm64.tar.gz bin/containers/cert-manager-controller-linux-s390x.tar.gz bin/containers/cert-manager-controller-linux-ppc64le.tar.gz bin/containers/cert-manager-controller-linux-arm.tar.gz: bin/containers/cert-manager-controller-linux-%.tar.gz: bin/server/controller-linux-% containers/Containerfil ...
+	$(eval TAG := cert-manager-controller-$*:$(RELEASE_VERSION))
+	$(eval BASE := BASE_IMAGE_$(notdir $<))
+	$(CTR) build --quiet \
```
`--quiet` is not available with `nerdctl` :( But it is available with `podman` and `docker`

maelvls replied:
nevermind, `--quiet` was recently added: https://github.com/containerd/nerdctl/pull/399

maelvls commented on make/Makefile:
```diff
@@ -0,0 +1,96 @@
+
+GOFLAGS=-ldflags '-w -s'
+# GOBUILDPROCS is passed to GOMAXPROCS when running go build
+GOBUILDPROCS=2
```
How can I set this to a bigger value?

SgtCoDFish replied:
You can!

```console
$ make -f make/Makefile -j8 GOBUILDPROCS=4 staged-release
```

That said, I'll actually change to this so that environment variables will also affect this variable:

```suggestion
GOBUILDPROCS ?= 2
```

Which would enable:

```console
$ export GOBUILDPROCS=4
$ make -f make/Makefile -j8 staged-release
```

maelvls replied:
Ahh, I did not know that passing a variable (`make VAR=foo`) was actually overriding variable assignments of the form `VAR=bar` in the makefile itself!

By the way, what is the rationale for not letting Go use the available cores? Is it to avoid overloading when using `-j8`?

SgtCoDFish replied:
> By the way, what is the rationale for not letting Go use the available cores? Is it to avoid overloading when using -j8?

Exactly, yeah. With `-jN` we can be running `go build` N times in parallel, and GOMAXPROCS defaults to the number of threads available on the system so it's just too much. This is a pain point with make; ideally I think we'd dynamically set GOMAXPROCS lower if running parallel but make doesn't really expose that information at runtime unfortunately (the value of `N` is _obtainable_ but it's hacky as hell).

It might make sense to have the default for `GOBUILDPROCS` be empty, so that the default behaviour is to use all cores when building individual binaries for testing on a laptop. We'd then manually set `GOBUILDPROCS` lower when running `make -f make/Makefile -jN staged-release` in cloudbuild.

maelvls commented on make/cmctl.mk:
```diff
@@ -0,0 +1,221 @@
+	mkdir -p $(TARDIR)
+	cp $< $(TARDIR)/cmctl.exe
+	cp bin/scratch/cert-manager.license $(TARDIR)/LICENSE
+	pushd $(TARDIR) && zip -r $(notdir $@) . && popd && mv $(TARDIR)/$(notdir $@) $@
```
`pushd` is not available on dash. dash is the default `/bin/sh` on Ubuntu. `/bin/sh` is the default shell used by `make`.

Do you think we could instead force the use of bash to avoid this? It is just a matter of adding:

```make
SHELL := /bin/bash
```

SgtCoDFish replied:
ah great idea! I actually meant to set the shell to bash and totally forgot.

I had _no idea_ that dash doesn't include pushd, either! Thanks for the learning :grin:

maelvls replied:
dash makes me sad 😢

maelvls commented:
The PR description is very good!! 

> After this PR merges, `cmrel stage` will essentially become "clone repo, run make staged-release, upload to a bucket".

Does it mean we are going to run `make` locally, and then upload `bin/release` a bucket manually with `gsutil`? 
Or are we going to be running some Cloud Build that will do that for us?

SgtCoDFish commented:
/test pull-cert-manager-e2e-v1-22

Seems clearly a flake

SgtCoDFish commented:
> The PR description is very good!!
Thanks :grin: 

> > After this PR merges, `cmrel stage` will essentially become "clone repo, run make staged-release, upload to a bucket".
> 
> Does it mean we are going to run `make` locally, and then upload `bin/release` a bucket manually with `gsutil`? If it is the case, shouldn't we have a make target `upload-release` for that? Or are we going to be running some Cloud Build that will do that for us?

I think that "upload-release" target should probably exist, yeah. I'll write it. All of this will run in Cloud Build for sure; that'll provide us all with centralized build logs, plus a stable but ephemeral build environment which can easily be defined in code.
 
> By the way, is there still a need for storing the staged artifacts to a bucket? Could we skip this step and upload the artifacts directly to the github release? And when we click "Publish", we have something that pushes these tarballs to quay.io?

I think storing them in a bucket is good; it's a central source of truth (and also a backup) of all release artifacts and that's valuable I think. The uploading logic could be moved into the main cert-manager repo longer term - I think all of `cmrel` could happily move into this repo.

SgtCoDFish commented:
/test pull-cert-manager-e2e-v1-22

SgtCoDFish commented:
The e2e test failure is unrelated and caused by https://github.com/jetstack/testing/pull/611

maelvls commented on make/Makefile:
```diff
@@ -0,0 +1,103 @@
+SHELL := /usr/bin/env bash
+.SHELLFLAGS := -uo pipefail -c
+
+SOURCES=$(shell find . -type f -name "*.go")
```
Here is a tiny tiny trick: with `=`, the makefile variable SOURCES is lazy-loaded every time it is used. Since it is used in about 30 different target (as part of the dependency list), the command `find . -type f -name "*.go"` is called 30 times.

```
$ time make -f make/Makefile help
make: *** No rule to make target 'help'.  Stop.
make -f make/Makefile help  0.23s user 0.25s system 101% cpu 0.477 total
```

If we instead go with a static variable declaration where the declaration expression is evaluated a single time, every `make` becomes  slightly faster (4 times faster for an existing target):

```
$ time make -f make/Makefile help
make: *** No rule to make target 'help'.  Stop.
make -f make/Makefile help  0.06s user 0.06s system 104% cpu 0.120 total
```

https://blog.jgc.org/2016/07/lazy-gnu-make-variables.html

SgtCoDFish replied:
Good catch, will make that change :+1:

SgtCoDFish commented:
Update: Working on adding the gcs upload + fixing containers atm, should have that finished soon

JoshVanL commented on containers/Containerfile.acmesolver:
```diff
@@ -0,0 +1,16 @@
+ARG BASE_IMAGE
```
Seems odd to have a dedicated root directory for this. Would this not be better to live under `/hack/containers`?

SgtCoDFish replied:
I think a dedicated root dir is reasonably justified here; our containers are our most important release artifacts and I think having our Containerfiles be fairly prominent is reasonable; plus there are tonnes of projects who have their Dockerfile in the root of the their repo.

I'd probably vote to move our helm chart template into `/helm` or to create `/release/{helm,containers}` in the longer term.

It doesn't matter though, I'll just move them into hack

JoshVanL commented on deploy/charts/cert-manager/Chart.template.yaml:
```diff
@@ -1,6 +1,5 @@
 apiVersion: v1
 name: cert-manager
-# The version and appVersion fields are set automatically by the release tool
 version: v0.1.0
```
Why has this line been removed? Is this no longer true?

SgtCoDFish replied:
Is is still true, but I didn't really see any value to the line. I'll just add it back

JoshVanL replied:
Yep, we got a non-insignificant number of issuers/PRs/comments around it before there was a comment (even when there was one!) https://github.com/jetstack/cert-manager/pull/2162

JoshVanL commented on hack/checkha.sh:
```diff
@@ -0,0 +1,27 @@
+#!/usr/bin/env bash
```
I was quite confused by the name `checksha.sh`, and thought it was a misspelling of `checksha`. Perhaps we can call it `check-hash.sh`?

JoshVanL replied:
This may also warrant living in `hack/util/check-hash.sh` to reduce bloat of the `hack` dir.

JoshVanL commented on hack/ha.sh:
```diff
@@ -0,0 +1,22 @@
+#!/usr/bin/env bash
```
Same re `hack/util`.

JoshVanL replied:
lol I have also realized now that "sh" has been omitted because that is the file extension.

Whilst cool, I think for completeness and reducing parsing time for humans, we should keep "sh" in the filename as well as the extension.

SgtCoDFish replied:
> lol I have also realized now that "sh" has been omitted because that is the file extension.

:cry: I'd never have even thought that there might be places where the extension isn't rendered, it's so rare that I see that nowadays

I'll rename!

JoshVanL commented on hack/latest-base-images.sh:
```diff
@@ -0,0 +1,38 @@
+# limitations under the License.
+
+set -eu -o pipefail
+
```
Can we add a comment here to what this script is doing?

JoshVanL commented on make/Makefile:
```diff
@@ -0,0 +1,103 @@
+include make/licenses.mk
+
+.PHONY: all
+all: binaries helm-chart static-manifests all-containers
+
+.PHONY: all-signed
+all-signed: all
+	$(MAKE) -f make/Makefile signed-artifacts
+
+.PHONY: binaries
+binaries: server-binaries cmctl kubectl-cert_manager
+
+.PHONY: signed-artifacts
+signed-artifacts: helm-chart-signature
+
+.PHONY: staged-release
+staged-release: all-signed release-manifests release-containers
+	$(MAKE) -f make/Makefile bin/release/metadata.json
```
Adding comments here for targets which we would expect a human to use would really help. I am also a fan of the Makefile magic used [here](https://github.com/cert-manager/approver-policy/blob/0c1601b1ee6cc533ff201bedd20c4b61e0444d94/Makefile#L28) which greatly improves UX on `make help`.

JoshVanL reviewed (changes requested):
Looking good from me. Just have a few things, most of them come down to file/dir structure/placement- slightly bikesheddy, but a good dir structure I think goes a long way. It is obviously a bit trickier atm with bazel still around and squatting all the good names.

Nice work :smile:

SgtCoDFish commented:
> Nice work

Thank you!

I think I've addressed all comments now; I didn't want to delve too deep into extending `make help` for _all_ possible useful invocations because they appear across so many different files ATM. There has to be a better way of surfacing targets such as `make tools`; probably those helpful .PHONY targets should all be defined in the main Makefile...

maelvls commented:
/lgtm

🔥🎄
